workflow/session.go

```diff
@@ -0,0 +1,127 @@
+// ActivityOptions. If none is specified, the default one will be used.
+//
+// CreationSession will fail in the following situations:
+//     1. The context passed in already contains a session which is still open
+//        (not closed and failed).
+//     2. All the workers are busy (number of sessions currently running on all the workers have reached
+//        MaxConCurrentSessionExecutionSize, which is specified when starting the workers) and session
+//        cannot be created within a specified timeout.
+//
+// If an activity is executed using the returned context, it's regarded as part of the
+// session. All activities within the same session will be executed by the same worker.
+// User still needs to handle the error returned when executing an activity. Session will
+// not be marked as failed if an activity within it returns an error. Only when the worker
+// executing the session is down, that session will be marked as failed. Executing an activity
+// within a failed session will return ErrSessionFailed immediately without scheduling that activity.
+//
+// If user wants to end a session since activity returns some error, use CompleteSession API below.
+// New session can be created if necessary to retry the whole session.
+//
+// Example:
+//    sessionCtx, err := CreateSession(ctx)
+//    if err != nil {
+//		    // Creation failed. Wrong ctx or too many outstanding sessions.
+//    }
+//    err = ExecuteActivity(sessionCtx, someActivityFunc, activityInput).Get(sessionCtx, nil)
+//    if err != nil {
+//        // Session has failed or activity itself failed.
+//    }
+//    ... // execute more activities using sessionCtx
+//    err = CompleteSession(sessionCtx)
+//    if err != nil {
+//        // Wrong ctx is used or failed to release session resource.
+//    }
+func CreateSession(ctx Context, sessionOptions *SessionOptions) (Context, error) {
+	return internal.CreateSession(ctx, sessionOptions)
+}
+
+// RecreateSession recreate a session based on the sessionInfo passed in. Activities executed within
+// the recreated session will be executed by the same worker as the previous session. CreateSessionForResourceID()
+// returns an error under the same situation as CreateSession() and has the same usage as CreateSession().
+// It will not check the state of the session described by the sessionInfo passed in, so user can recreate
+// a session based on a failed or completed session.
+//
+// The main usage of RecreateSession is for long sessions that are splited into multiple runs. At the end of
+// one run, complete the current session, get recreateSessionParams from sessionInfo and pass the parameter to
+// next run. In the new run, the session can be recreated using the parameter.
+func RecreateSession(ctx Context, params *RecreateSessionParams, sessionOptions *SessionOptions) (Context, error) {
+	return internal.RecreateSession(ctx, params, sessionOptions)
+}
```

Comment:
Hide RecreateSession as it is not possible to recreate yet using the public APIs.

Reply:
I think it's possible to recreate a session. The RecreateSessionParams can be got by GetSessionInfo(ctx).GetRecreateParams().